hail/python/hail/methods/relatedness/pc_relate_bm.py

```diff
@@ -0,0 +1,536 @@
+    """
+    assert (0.0 <= min_individual_maf <= 1.0), \
+        f"invalid argument: min_individual_maf={min_individual_maf}. " \
+        f"Must have min_individual_maf on interval [0.0, 1.0]."
+    mt = matrix_table_source('pc_relate_bm/call_expr', call_expr)
+
+    if k and scores_expr is None:
+        _, scores, _ = _hwe_normalized_blanczos(call_expr, k, compute_loadings=False, q_iterations=10)
+        scores_expr = scores[mt.col_key].scores
+    elif not k and scores_expr is not None:
+        analyze('pc_relate_bm/scores_expr', scores_expr, mt._col_indices)
+    elif k and scores_expr is not None:
+        raise ValueError("pc_relate_bm: exactly one of 'k' and 'scores_expr' "
+                         "must be set, found both")
+    else:
+        raise ValueError("pc_relate_bm: exactly one of 'k' and 'scores_expr' "
+                         "must be set, found neither")
+
+    scores_table = mt.select_cols(__scores=scores_expr) \
+        .key_cols_by().select_cols('__scores').cols()
+
+    n_missing = scores_table.aggregate(agg.count_where(hl.is_missing(scores_table.__scores)))
+    if n_missing > 0:
+        raise ValueError(f'Found {n_missing} columns with missing scores array.')
+
+    mt = mt.select_entries(__gt=call_expr.n_alt_alleles()).unfilter_entries()
+    mt = mt.annotate_rows(__mean_gt=agg.mean(mt.__gt))
+    mean_imputed_gt = hl.or_else(hl.float64(mt.__gt), mt.__mean_gt)
+
+    if not block_size:
+        block_size = BlockMatrix.default_block_size()
+
+    g = BlockMatrix.from_entry_expr(mean_imputed_gt, block_size=block_size)
+    pc_scores = hl.nd.array(scores_table.collect(_localize=False).map(lambda x: x.__scores))
+
+    # g.shape is (n_variants, n_samples), pc_scores.shape is (n_samples, k)
+    sqrt_n_samples = hl.nd.array([hl.sqrt(g.shape[1])])
+
+    # Recover singular values, S0, as vector of column norms of pc_scores
+    S0 = (pc_scores ** hl.int32(2)).sum(0).map(lambda x: hl.sqrt(x))
+    # We want sqrt(n_samples) as first entry in S, for intercept in beta
+    S = hl.nd.hstack((sqrt_n_samples, S0))._persist()
+
+    # Recover V from pc_scores with inv(S0)
+    V0 = (pc_scores * (1 / S0))._persist()
+    # First column in V needs all entries as 1/sqrt(n_samples), for intercept in beta
+    ones_normalized = hl.nd.full((V0.shape[0], 1), (1 / S[0]))
+    V = hl.nd.hstack((ones_normalized, V0))
+
+    beta = (BlockMatrix.from_ndarray(((1 / S) * V).T, block_size=block_size) @ g.T) \
+        .checkpoint(new_temp_file("pc_relate_bm/beta", "bm"))
+    mu = (0.5 * (BlockMatrix.from_ndarray(V * S, block_size=block_size) @ beta).T) \
+        .checkpoint(new_temp_file("pc_relate_bm/pre-mu", "bm"))
+
+    # Define NaN to use instead of missing, otherwise cannot go back to block matrix
+    nan = hl.float64(float("NaN"))
+
+    # Replace bad entries in g and pre_mu with NaNs
+    g = g._map_dense(lambda x: hl.if_else(_bad_gt(x), nan, x)).checkpoint(new_temp_file("pc_relate_bm/g", "bm"))
```

Comment:
Trying to follow this. 

`g` comes from `mean_imputed_gt`, which has fields that are floats derived from mean imputing data. Do we intend to throw those mean imputed values out if they don't land exactly on 0.0, 1.0, or 2.0?

Reply:
This is a good point, I had to take a look at the paper again to refresh my memory. In the GRM definition in the beginning (equation 1) it says

> where S_ij is the subset of SNPs for which individuals i and j both have non-missing genotype data

so maybe it's more appropriate to skip the mean imputation all together. Otherwise we could potentially end up keeping mean imputed values that happen to be exactly 0, 1, or 2, as you point out.

Reply:
Based on your quote from the paper, I think it would be correct to skip the mean imputing.

Reply:
Probably should just introduce nans at the time where you were doing mean imputing, and then you don't need this `_bad_gt` method at all

Reply:
Cool, I'll make that change